Perform check_config service in current process

kellerza opened the pull request:
## Description:
Run check_config directly instead of executing the script in a subprocess

## Checklist:
  - [x] The code change is tested and works locally.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

kellerza committed:
Perform check_config service in current process

balloob commented on homeassistant/config.py:
```diff
@@ -662,22 +660,16 @@ def async_process_component_config(hass, config, domain):
-    return None
+    from homeassistant.scripts.check_config import check_ha_config_file
+
+    def _load_hass_yaml_config():
```
There is no need fo this, just add the 2 methods to async_add_job:

```
res = await hass.async_add_job(check_ha_config_file, hass.config.config_dir)
```

balloob commented on tests/test_config.py:
```diff
@@ -514,35 +514,32 @@ def test_discovering_configuration_auto_detect_fails(self, mock_detect,
-    @mock.patch('asyncio.create_subprocess_exec')
-    def test_check_ha_config_file_correct(self, mock_create):
+    @mock.patch('os.path.isfile', return_value=True)
+    def test_check_ha_config_file_correct(self, mock_file):
```
We already have tests for `check_ha_config_file` so we should stub that out here and just have it return a dummy error log.

balloob replied:
We should try to do 0 I/O inside our tests. I/O in tests make them unreliable and slow (especially on Travis).

kellerza committed:
feedback

kellerza commented:
Not attempted in this PR, but the notification when you call homeassistant/restart can do with some improving:

If successful, you get: `Failed to call service homeassistant/restarts.` and it restarts.

If failed, you get `Service homeassistant/restart called.` and nothing else --> check the log and you get a list of the errors

balloob commented:
@kellerza do you remember why we did this?

I want to reverse this, as the core should not have a dependency on a script. Nor is this script something that can currently run inside the HASS process, because it mocks out a lot of things / clears things.

kellerza commented:
Someone performed a hass check_config from the gui “restart” request, which was terribly slow and this started this (although forking does not interfere with the current process)

Technically we don't need to mock anything for a quick config check and I have already split the check_config script in 2 parts. One for a quick check- almost no mocks(target is none) and then a more advanced version that mocks secrets and some parts of the loader to print info and secret usage

If you want I can complete this work and split check_config to homeassistant.helpers.check_config.py?

balloob commented:
That would be great. I was looking at it last night, and indeed a lot of code can be re-used. 

Still not sure how you would do the secrets part though? Maybe we can leave that for the very last PR.
